### ISSUE TITLE
[CUDA] illegal memory access on max_pool2d/max_pool3d with integer-overflow dilation

### ISSUE THREAD

**Issue by kokol16** [pytorch/pytorch#189664]
### 🐛 Describe the bug

The OOB is caused by dilation=INT32_MAX overflowing the 32-bit loop counter in the CUDA max-pooling kernel.
```python
# Bug: max_pool_forward_nchw 32-bit index overflow on huge dilation
import torch
print("torch:", torch.__version__)
torch.cuda.init()
x = torch.zeros(5, 10, 24, 24, device='cuda')
torch.nn.functional.max_pool2d(x, kernel_size=1, stride=None, padding=0,
                               dilation=2147483647, ceil_mode=True)
torch.cuda.synchronize()
print("done (no error)")
```

```text
========= COMPUTE-SANITIZER
========= Invalid __global__ read of size 4 bytes
=========     at void at::native::<unnamed>::max_pool_forward_nchw<float, int>(T2, const T1 *, long, long, long, int, int, int, int, int, int, int, int, int, int, T1 *, long *)+0xca0
=========     by thread (1,0,0) in block (0,0,0)
=========     Address 0x7fd6c5e00000 is out of bounds
=========     and is 8589934592 bytes before the nearest allocation at 0x7fd8c5e00000 of size 2097152 bytes
```

### Versions

Collecting environment information...

/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.8.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  104
On-line CPU(s) list:                     0-103
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Gold 5320 CPU @ 2.20GHz
CPU family:                              6
Model:                                   106
Thread(s) per core:                      2
Core(s) per socket:                      26
Socket(s):                               2
Stepping:                                6
CPU(s) scaling MHz:                      39%
CPU max MHz:                             3400.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4400.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               2.4 MiB (52 instances)
L1i cache:                               1.6 MiB (52 instances)
L2 cache:                                65 MiB (52 instances)
L3 cache:                                78 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-25,52-77
NUMA node1 CPU(s):                       26-51,78-103
Vulnerability Gather data sampling:      Mitigation; Microcode
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.85
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.29.3
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.13.0+cu126
[pip3] triton==3.7.1
[conda] Could not collect

cc @ptrblck @msaroufim @eqy @tinglvv @nWEIdia @malfet @mikaylagawarecki

**Comment by MDSALMANSHAMS** [comment]
With `dilation=2147483647` the effective pooling-window extent overflows the 32-bit index math in `max_pool_forward_nchw`, so the window start/end land outside the input and the kernel reads OOB. A validation guard rejecting a kernel/stride/dilation combination whose effective window can't fit the (padded) input — or that overflows the index type — would turn this into a clear error at the ATen level before launch. Would maintainers welcome a PR adding that check? If so, could this be marked `actionable`?

**Comment by MDSALMANSHAMS** [comment]
Thanks for consolidating these, @malfet — that matches what I found on the 3d side.

I'll put up a fix with regression tests covering both kernels. Plan:

- Guard in the shared shape checks in `aten/src/ATen/native/Pool.h` (`pool2d_shape_check` / `pool3d_shape_check`), which today only validate `dilation > 0` with no upper bound.
- Reject any kernel/stride/dilation/padding combination whose effective window extent overflows the 32-bit index math in `max_pool_forward_nchw` (`DilatedMaxPool2d.cu`) and `max_pool3d_with_indices_single_out_frame` (`DilatedMaxPool3d.cu`), so it fails with a clear `TORCH_CHECK` at the ATen level instead of an out-of-bounds read inside the kernel.
- Host-side only — kernel/stride/dilation are `IntArrayRef`, so this is metadata arithmetic with no device-tensor value reads and therefore no sync.
- Regression tests for both 2d and 3d, verified under compute-sanitizer.

Could you mark this `actionable` so it's greenlit before I open the PR? Same shape as the fix that landed in #190005.

**Comment by malfet** [comment]
Sure, though I want it fixed in hacked agnostic way: I.e. cpu should probably return the same error, to be checked via OpInfo error inputs

**Comment by CodersAcademy006** [comment]
@MDSALMANSHAMS your plan covers the CUDA side, and @malfet's ask for a backend-agnostic error looks like it needs a second piece your outline does not mention: CPU currently does not raise for the same input, so a guard placed only in the CUDA path would leave the two backends disagreeing. Happy to take that half rather than duplicate yours, meaning the CPU-side error and the `error_inputs` entries in `OpInfo` for `max_pool2d`/`max_pool3d` so the parity is enforced by the shared test rather than by two separate regression tests.

If you would rather carry the whole thing in one PR, that is fine too and I will stay out of it, just say so. Otherwise I will wait until yours is up so I can build the `error_inputs` on top of whatever `TORCH_CHECK` message you settle on, which keeps the two backends emitting the same string.

**Comment by MDSALMANSHAMS** [comment]
Thanks for coordinating rather than just opening something in parallel — appreciated. I'll carry the whole thing in one PR, so please do stay out of this one.

On the concern about the two backends disagreeing: the guard isn't going in the CUDA path. It goes in `pool2d_shape_check` / `pool3d_shape_check` in `aten/src/ATen/native/Pool.h`, which both backends already share:

- **2d:** `pool2d_shape_check` (`Pool.h:122`) is called from `TORCH_META_FUNC(max_pool2d_with_indices)` at `aten/src/ATen/native/DilatedMaxPool2d.cpp:68` — a structured-kernel meta function, so it runs on CPU and CUDA dispatch alike.
- **3d:** `pool3d_shape_check` is called from `DilatedMaxPool3d.cpp:77` (CPU) and `cuda/DilatedMaxPool3d.cu:336` (CUDA) — two call sites, one shared function. `mps/operations/Pooling.mm` goes through the same header.

So a single `TORCH_CHECK` there makes both backends emit the identical string by construction, and there's no separate CPU half to split off. Today that function validates only `dilation > 0` (`Pool.h:138`) with no upper bound, which is precisely the gap — `dilation=2147483647` survives the `checked_convert<int>` narrowing and then overflows the window/index arithmetic downstream.

Scope I'll cover, per @malfet's ask for a device-agnostic fix:

- upper-bound guard in both shape checks in `Pool.h`, rejecting any kernel/stride/dilation/padding whose effective window overflows the 32-bit index math
- `error_inputs` on the `max_pool2d` / `max_pool3d` OpInfos, so CPU/CUDA parity is enforced by the shared test rather than by two separate regression tests
- mirror the check into `torch/_meta_registrations.py` if the meta path needs it to agree

I'll verify under `compute-sanitizer` on a CUDA box and have the PR up in the next few days.

**Comment by malfet** [comment]
@MDSALMANSHAMS you don't really need to verify with compute sanitizer: if check can be made at the meta layer, it's more than sufficient and will also prevent a follow up issue about eager and compile disagreeing on how to treat those 

@CodersAcademy006 as problem stated it can not be done in two PR, as check must be uniform for CPU/CUDA/MPS...

**Comment by MDSALMANSHAMS** [comment]
Perfect, thanks - I'll put the check at the meta layer then, with `error_inputs` on the `max_pool2d`/`max_pool3d` OpInfos. PR shortly.

**Comment by malfet** [comment]
Duplicate of #195547. This is one instance of a class of missing shape/size checks that lead to out-of-bounds tensor access — here surfaced by compute-sanitizer on a CUDA kernel rather than by ASAN on CPU, but the same missing bounds contract. The class is now tracked under the umbrella issue #195547, which also explains why per-operator TORCH_CHECK PRs are not the fix being sought. Please follow #195547.
